### PR TITLE
Define structs and functionize request parsing

### PR DESCRIPTION
**Pull request by humblec**
Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

**Describe what this PR does**
 This PR functionize parsing request and fill data on subjected structs. The plan is to use these structs for storing gd2 specific values and also CSI driver specific values across the source or servers.

**Is there anything that requires special attention?**
Do you have any questions? Did you do something clever?

**Related issues:**
Mention any github issues relevant to this PR

### CODE REVIEW

File: pkg/glusterfs/controllerserver.go [hunk 1]
```diff
@@ -56,16 +97,21 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
...
-	glusterURLPort := req.GetParameters()["glusterurlport"]
-	glusterUser := req.GetParameters()["glusteruser"]
-	glusterUserSecret := req.GetParameters()["glusterusersecret"]
+	// parse the request.
+	parseResp, parseErr := cs.ParseRequest(req)
+	if parseErr != nil {
+		return nil, status.Error(codes.InvalidArgument, "failed to parse request")
+	}
+
+	if parseResp != nil && parseResp.gdVolReq != nil {
```
Comment (JohnStrunk) [inline pkg/glusterfs/controllerserver.go]: This is covered in the contract of ParseRequest

File: pkg/glusterfs/controllerserver.go [hunk 2]
```diff
@@ -56,16 +97,21 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
...
+		if len(parseResp.gdVolReq.Name) == 0 {
```
Comment (JohnStrunk) [inline pkg/glusterfs/controllerserver.go]: Can this be covered in ParseRequest or are there situations where an empty volumename are acceptable?

File: pkg/glusterfs/controllerserver.go [hunk 3]
```diff
@@ -56,16 +97,21 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
...
+			return nil, status.Error(codes.InvalidArgument, "volumename is nil")
+		}
+	} else {
+		return nil, status.Error(codes.InvalidArgument, "parse response in nil")
+	}
 
-	glog.V(3).Infof("Request fields:[ %v %v %v %v %v %v]", glusterVol, glusterServer, glusterURL, glusterURLPort, glusterUser, glusterUserSecret)
```
Comment (JohnStrunk) [inline pkg/glusterfs/controllerserver.go]: Did this go away on purpose?

File: pkg/glusterfs/controllerserver.go [hunk 4]
```diff
@@ -29,6 +29,47 @@ type ControllerServer struct {
...
+	GlusterUser      string
+	GlusterUserToken string
+	GlusterVersion   string
+	CsiDrvName       string
+	CsiDrvVersion    string
+	CompMatrix       map[string]string
+}
+
+// RequestConfig is the final struct after parsing request and CSI driver specific input
+type RequestConfig struct {
```
Comment (JohnStrunk) [inline pkg/glusterfs/controllerserver.go]: Why not embed CsiDrvParam & api.VolCreateReq?

File: pkg/glusterfs/controllerserver.go [hunk 5]
```diff
@@ -29,6 +29,47 @@ type ControllerServer struct {
...
+	gdVolReq *api.VolCreateReq
+	csiConf  *CsiDrvParam
+}
+
+func (cs *ControllerServer) ParseRequest(req *csi.CreateVolumeRequest) (*RequestConfig, error) {
```
Comment (JohnStrunk) [inline pkg/glusterfs/controllerserver.go]: The names `RequestConfig` and `ParseRequest` don't mention they are specific to creating a volume. Won't we end up w/ a pair of these for each type of request?

File: pkg/glusterfs/controllerserver.go [hunk 6]
```diff
@@ -29,6 +29,47 @@ type ControllerServer struct {
...
+//              version matrix between CSI driver and gluster cluster. All these fields are optional and can be used if needed.
+
+type CsiDrvParam struct {
+	GlusterCluster   string
+	GlusterUser      string
+	GlusterUserToken string
+	GlusterVersion   string
+	CsiDrvName       string
+	CsiDrvVersion    string
+	CompMatrix       map[string]string
```
[the remark this answers is not in the document]
Comment (humblec) [reply pkg/glusterfs/controllerserver.go]: The `compMatrix` intended to keep the supportability matrix between supported gluster server version and CSI driver version, which is purely internal. There is also a possibility of exposing an API call in CSI driver which can be queried by an external client ( for ex: operator) not CO or other CSI specific drivers. This matrix if exposed,  allows an external client to take some actions based on this field and external client is also free to query directly gluster and get the running version . But, I can drop this field from this PR, in other way,  keeping it in internal struct should also not harm. @JohnStrunk let me know your thought.

File: pkg/glusterfs/controllerserver.go (hunk 3 above)
Comment (humblec) [reply pkg/glusterfs/controllerserver.go]: Yes, I removed it deliberately. it was rather a debugging entry.

File: pkg/glusterfs/controllerserver.go (hunk 5 above)
Comment (humblec) [reply pkg/glusterfs/controllerserver.go]: Let me find a better name then :)

File: pkg/glusterfs/controllerserver.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (JohnStrunk) [reply pkg/glusterfs/controllerserver.go]: Please remove it until we have a concrete use for it.

File: pkg/glusterfs/controllerserver.go (hunk 1 above)
Comment (humblec) [reply pkg/glusterfs/controllerserver.go]: ParseReq check the input request and here we check the response before deferencing it.

File: pkg/glusterfs/controllerserver.go (hunk 2 above)
Comment (humblec) [reply pkg/glusterfs/controllerserver.go]: its not required, removing it.

File: pkg/glusterfs/controllerserver.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (humblec) [reply pkg/glusterfs/controllerserver.go]: Done.

File: pkg/glusterfs/controllerserver.go (hunk 4 above)
Comment (humblec) [reply pkg/glusterfs/controllerserver.go]: api.VolCreateReq is the client API of GD2. So, embedding is not possible in it.

**Comment by humblec** [comment]
@JohnStrunk comments are addressed, PTAL .

**Comment by humblec** [comment]
ok to test

**Comment by humblec** [comment]
@Madhu-1 any comments ?

**Comment by Madhu-1** [comment]
@humblec Good to merge :+1:

**Comment by humblec** [comment]
As I have approval from other maintainers of this repo, I am merging this PR.

**Pull request merged**